### PR TITLE
Move sorting logic to artifact

### PR DESCRIPTION
**Pull request by runesoerensen**
[WIP - don't review just yet]

This PR builds on https://github.com/heroku/buildpacks-go/pull/245 and:

* Moves artifact sorting logic to the `Artifact` and ensures that artifacts are explicitly sorted - more details here: https://github.com/heroku/buildpacks-go/pull/246/commits/0eb2de9fd88bcb081ec831afa7132b7ef49071ee.
* Retains existing functionality by also order by the `arch` (for ease of review - we may want to change this later, or add more sorting parameters). More info here: https://github.com/heroku/buildpacks-go/pull/246/commits/4ad59291875fcce00a8a75f7493fb92fea08d571
* Continues to allow consumers of the shared inventory code to implement different versioning ordering schemes appropriate for their version schemes and requirements.

### CODE REVIEW

File: common/inventory-utils/src/inv.rs [hunk 1]
```diff
@@ -30,17 +30,29 @@ where
     pub checksum: Checksum,
 }
 
+impl<V: Version> Ord for Artifact<V> {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
```
Comment (colincasey) [inline common/inventory-utils/src/inv.rs:34]: Add some unit tests here.

File: common/inventory-utils/src/inv.rs [hunk 2]
```diff
@@ -30,17 +30,29 @@ where
...
+        (&self.version, &self.arch).cmp(&(&other.version, &other.arch))
+    }
+}
+
+impl<V: Version> PartialOrd for Artifact<V> {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
```
Comment (colincasey) [inline common/inventory-utils/src/inv.rs:40]: Add some unit tests here.

File: common/go-utils/src/bin/diff_inventory.rs [hunk 3]
```diff
@@ -39,7 +39,8 @@ fn main() {
     .filter(|(_, artifact_diff)| !artifact_diff.is_empty())
     .for_each(|(action, artifacts)| {
         let mut list: Vec<&Artifact<GoVersion>> = artifacts.iter().collect();
-        list.sort_by_key(|a| &a.version);
+        list.sort();
+        list.reverse();
```
Comment (colincasey) [inline common/go-utils/src/bin/diff_inventory.rs:43]: Does this mean the default sorting is version ascending (i.e.; oldest to newest)? Shouldn't it be the opposite so you don't need to `reverse()` it?

**Review by joshwlewis — CHANGES_REQUESTED** [review]
Implementation looks good, but I agree that it needs unit tests.

**Pull request closed without merging**